src/workerd/api/encoding.c++

```diff
@@ -491,34 +498,255 @@ jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
+      // Don't even bother checking any more, it's all going to fit.  Hitting
+      // this halfway through is also a good reason to limit the CHUNK size.
+      return length;
+    }
+    size_t likelyToFit = kj::min(static_cast<size_t>(spaceRemaining / expansion), CHUNK);
+    size_t fitEstimate = kj::max(1, kj::max(guaranteedToFit, likelyToFit));
+    size_t chunkSize = kj::min(remainingInput, fitEstimate);
+    if (chunkSize == 1) break;  // Not worth running this complicated stuff one char at a time.
+    // No div-by-zero because remainingInput and fitEstimate are at least 1.
+    KJ_DASSERT(chunkSize >= 1);
+
+    size_t chunkUtf8Len;
+    if constexpr (UTF16) {
+      chunkUtf8Len = simdutf::utf8_length_from_utf16_with_replacement(data + pos, chunkSize).count;
+    } else {
+      chunkUtf8Len = simdutf::utf8_length_from_latin1(data + pos, chunkSize);
+    }
+
+    if (utf8Accumulated + chunkUtf8Len > bufferSize) {
+      // Our chosen chunk didn't fit in the rest of the output buffer.
+      KJ_DASSERT(chunkSize > guaranteedToFit);
+      // Since it didn't fit we adjust our expansion guess upwards.
+      expansion = kj::max(expansion * 1.1, (chunkUtf8Len * 1.1) / chunkSize);
+    } else {
+      // Use successful length calculation to adjust our expansion estimate.
+      expansion = kj::max(1.15, (chunkUtf8Len * 1.1) / chunkSize);
+      pos += chunkSize;
+      utf8Accumulated += chunkUtf8Len;
+    }
+  }
+  // Do the last few code units in a simpler way.
+  while (pos < length && utf8Accumulated < bufferSize) {
+    size_t extra = simpleUtfEncodingLength(data[pos]);
+    if (utf8Accumulated + extra > bufferSize) break;
+    pos++;
+    utf8Accumulated += extra;
+  }
+  if (UTF16 && pos != 0 && pos != length && isSurrogatePair(data[pos - 1], data[pos])) {
+    // We ended on a leading surrogate which has a matching trailing surrogate in the next
+    // position.  In order to make progress when the bufferSize is tiny we try to include it.
+    if (utf8Accumulated < bufferSize) {
+      pos++;  // We had one more byte, so we can include the pair, UTF-8 encoding 3->4.
+    } else {
+      pos--;  // Don't chop the pair in half.
+    }
+  }
+  return pos;
+}
+
 }  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+namespace test {
+
+size_t bestFit(const char* str, size_t bufferSize) {
+  return findBestFit(str, strlen(str), bufferSize);
+}
+
+size_t bestFit(const char16_t* str, size_t bufferSize) {
+  size_t length = 0;
+  while (str[length] != 0) length++;
+  return findBestFit(str, length, bufferSize);
 }
 
+}  // namespace test
```

Comment:
A comment in here that these are used in encoding-test.c++ would be helpful. It's easy to miss since they aren't used anywhere in this file.